app/graph.go

```diff
@@ -0,0 +1,193 @@
+func (dag Dag) Order() int {
+	return len(dag.NodeMap)
+}
+
+// Visit calls the do function for each neighbor w of vertex v, used by the graph acyclic validator
+func (dag Dag) Visit(v int, do func(w int, c int64) (skip bool)) (aborted bool) {
+	for _, edge := range dag.EdgesMap[v] {
+		// just have cost as zero because we only need for acyclic validation purposes
+		if do(edge.ToNodeID, 0) {
+			return true
+		}
+	}
+	return false
+}
+
+func NewDag() Dag {
+	return Dag{
+		NodeMap:      make(map[int]DagNode),
+		EdgesMap:     make(map[int][]DagEdge),
+		AccessOpsMap: make(map[acltypes.AccessOperation][]int),
+		TxIndexMap:   make(map[int]int),
+		NextID:       0,
+	}
+}
+
+func (dag *Dag) AddNode(txIndex int, accessOp acltypes.AccessOperation) DagNode {
+	dagNode := DagNode{
+		NodeID:          dag.NextID,
+		TxIndex:         txIndex,
+		AccessOperation: accessOp,
+	}
+	dag.NodeMap[dag.NextID] = dagNode
+	dag.NextID++
+	return dagNode
+}
+
+func (dag *Dag) AddEdge(fromIndex int, toIndex int, accessOp *acltypes.AccessOperation) *DagEdge {
+	// no-ops if the from or to node doesn't exist
+	if _, ok := dag.NodeMap[fromIndex]; !ok {
+		return nil
+	}
+	if _, ok := dag.NodeMap[toIndex]; !ok {
+		return nil
+	}
+	newEdge := DagEdge{fromIndex, toIndex, accessOp}
+	dag.EdgesMap[fromIndex] = append(dag.EdgesMap[fromIndex], newEdge)
+	return &newEdge
+}
+
+func (dag *Dag) AddNodeBuildDependency(ctx sdk.Context, txIndex int, accessOp acltypes.AccessOperation) {
+	dagNode := dag.AddNode(txIndex, accessOp)
+	// if in TxIndexMap, make an edge from the previous node index
+	if lastTxNodeID, ok := dag.TxIndexMap[txIndex]; ok {
+		// add an edge with no access op
+		dag.AddEdge(lastTxNodeID, dagNode.NodeID, nil)
+	}
+	// update tx index map
+	dag.TxIndexMap[txIndex] = dagNode.NodeID
+	// if the blocking access ops are in access ops map, make an edge
+	switch accessOp.AccessType {
+	case acltypes.AccessType_READ:
+		// if we need to do a read, we need latest write as a dependency
+		// TODO: replace hardcoded access op dependencies with helper that generates (and also generates superseding resources too eg. Resource.ALL is blocking for Resource.KV)
+		writeAccessOp := acltypes.AccessOperation{
+			AccessType:         acltypes.AccessType_WRITE,
+			ResourceType:       accessOp.GetResourceType(),
+			IdentifierTemplate: accessOp.GetIdentifierTemplate(),
+		}
+		if writeNodeIDs, ok := dag.AccessOpsMap[writeAccessOp]; ok {
+			for _, wn := range writeNodeIDs {
+				writeNode := dag.NodeMap[wn]
+				// if accessOp exists already (and from a previous transaction), we need to define a dependency on the previous message (and make a edge between the two)
+				// if from a previous transaction, we need to create an edge
+				if writeNode.TxIndex < dagNode.TxIndex {
+					lastTxNode := dag.NodeMap[dag.TxIndexMap[writeNode.TxIndex]]
+					dag.AddEdge(lastTxNode.NodeID, dagNode.NodeID, &writeAccessOp)
+				}
+			}
+		}
+	case acltypes.AccessType_WRITE:
```

Comment:
I believe there's also `AccessType_ANY`, though I think that might be similar to Accesstype_WRITE since it should just block on previous reads and writes

Reply:
wait I double checked and it has access type UNKNOWN. Do we want to handle it as if it's ANY (and/or refactor to ANY?)

Reply:
Ah, I think they should be the same thing. If you feel that ANY is more clear then I can make a PR to change it